File: javascript/socketd-examples/demo01-browser-websocket-npm/js/socketd/transport/core/Processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProcessorDefault = void 0;
const Listener_1 = require("./Listener");
const Constants_1 = require("./Constants");
const SocketdException_1 = require("../../exception/SocketdException");
const HandshakeDefault_1 = require("./HandshakeDefault");
class ProcessorDefault {
    constructor() {
        this._listener = new Listener_1.SimpleListener();
    }
    setListener(listener) {
        if (listener != null) {
            this._listener = listener;
        }
    }
    onReceive(channel, frame) {
        if (channel.getConfig().clientMode()) {
            console.debug("C-REV:" + frame);
        }
        else {
            console.debug("S-REV:" + frame);
        }
        if (frame.flag() == Constants_1.Flags.Connect) {
            channel.setHandshake(new HandshakeDefault_1.HandshakeDefault(frame.message()));
            channel.onOpenFuture((r, err) => {
                if (r && channel.isValid()) {
                    //如果还有效，则发送链接确认
                    try {
                        channel.sendConnack(frame.message()); //->Connack
                    }
                    catch (err) {
                        this.onError(channel, err);
                    }
                }
            });
            this.onOpen(channel);
        }
        else if (frame.flag() == Constants_1.Flags.Connack) {
            //if client
            channel.setHandshake(new HandshakeDefault_1.HandshakeDefault(frame.message()));
            this.onOpen(channel);
        }
        else {
            if (channel.getHandshake() == null) {
                channel.close(Constants_1.Constants.CLOSE1_PROTOCOL);
                if (frame.flag() == Constants_1.Flags.Close) {
                    //说明握手失败了
                    throw new SocketdException_1.SocketdConnectionException("Connection request was rejected");
                }
                console.warn(`${channel.getConfig().getRoleName()} channel handshake is null, sessionId=${channel.getSession().sessionId()}`);
                return;
            }
            try {
                switch (frame.flag()) {
                    case Constants_1.Flags.Ping: {
                        channel.sendPong();
                        break;
                    }
                    case Constants_1.Flags.Pong: {
                        break;
                    }
                    case Constants_1.Flags.Close: {
                        //关闭通道
                        channel.close(Constants_1.Constants.CLOSE1_PROTOCOL);
                        this.onCloseInternal(channel);
                        break;
                    }
                    case Constants_1.Flags.Alarm: {
                        //结束流，并异常通知
                        const exception = new SocketdException_1.SocketdAlarmException(frame.message());
                        const stream = channel.getConfig().getStreamManger().getStream(frame.message().sid());
                        if (stream == null) {
                            this.onError(channel, exception);
                        }
                        else {
                            channel.getConfig().getStreamManger().removeStream(frame.message().sid());
                            stream.onError(exception);
                        }
                        break;
                    }
                    case Constants_1.Flags.Message:
                    case Constants_1.Flags.Request:
                    case Constants_1.Flags.Subscribe: {
                        this.onReceiveDo(channel, frame, false);
                        break;
                    }
                    case Constants_1.Flags.Reply:
                    case Constants_1.Flags.ReplyEnd: {
                        this.onReceiveDo(channel, frame, true);
                        break;
                    }
                    default: {
                        channel.close(Constants_1.Constants.CLOSE2_PROTOCOL_ILLEGAL);
                        this.onCloseInternal(channel);
                    }
                }
            }
            catch (e) {
                this.onError(channel, e);
            }
        }
    }
    onReceiveDo(channel, frame, isReply) {
        //如果启用了聚合!
        if (channel.getConfig().getFragmentHandler().aggrEnable()) {
            //尝试聚合分片处理
            const fragmentIdxStr = frame.message().meta(Constants_1.EntityMetas.META_DATA_FRAGMENT_IDX);
            if (fragmentIdxStr != null) {
                //解析分片索引
                const index = parseInt(fragmentIdxStr);
                const frameNew = channel.getConfig().getFragmentHandler().aggrFragment(channel, index, frame.message());
                if (frameNew == null) {
                    return;
                }
                else {
                    frame = frameNew;
                }
            }
        }
        //执行接收处理
        if (isReply) {
            channel.retrieve(frame);
        }
        else {
            this.onMessage(channel, frame.message());
        }
    }
    onOpen(channel) {
        try {
            this._listener.onOpen(channel.getSession());
            channel.doOpenFuture(true, null);
        }
        catch (e) {
            console.warn(`${channel.getConfig().getRoleName()} channel listener onOpen error`, e);
            channel.doOpenFuture(false, e);
        }
    }
    onMessage(channel, message) {
        try {
            this._listener.onMessage(channel.getSession(), message);
        }
        catch (e) {
            console.warn(`${channel.getConfig().getRoleName()} channel listener onMessage error`, e);
            this.onError(channel, e);
        }
    }
    onClose(channel) {
        if (channel.isClosed() == 0) {
            this.onCloseInternal(channel);
        }
    }
    onCloseInternal(channel) {
        this._listener.onClose(channel.getSession());
    }
    onError(channel, error) {
        this._listener.onError(channel.getSession(), error);
    }
}
exports.ProcessorDefault = ProcessorDefault;

File: go/socketd/transport/core/impl/processor.go
package impl

import (
	"errors"
	"fmt"
	"log/slog"
	"strconv"

	"socketd/transport/core/constant"
	"socketd/transport/stream"

	"socketd/transport/core"
	"socketd/transport/core/message"
)

type Processor struct {
	listener core.Listener
}

func NewProcessor() *Processor {
	return &Processor{}
}

func (p *Processor) SetListener(listener core.Listener) {
	p.listener = listener
}

func (p *Processor) OnReceive(channel core.ChannelInternal, frame *message.Frame) error {
	slog.Debug("OnReceive", "frame", frame)

	switch frame.Flag {
	case constant.FrameConnect:
		var handshake = core.NewHandshake(frame.Message)
		channel.SetHandshake(handshake)

		channel.OnOpenFuture(func(r bool, e error) {
			if channel.IsValid() {
				if e != nil {
					channel.Close(constant.CLOSE3_ERROR)
					p.OnCloseInternal(channel)
				} else {
					var err = channel.SendConnectAck(frame.Message)
					if err != nil {
						p.OnError(channel, err)
					}
				}
			}
		})
		p.OnOpen(channel)
	case constant.FrameConnack:
		var handshake = core.NewHandshake(frame.Message)
		channel.SetHandshake(handshake)

		p.OnOpen(channel)
	default:
		if channel.GetHandshake() == nil {
			channel.Close(constant.CLOSE1_PROTOCOL)

			// 握手失败
			if frame.Flag == constant.FrameClose {
				return errors.New("connection request was rejected")
			}

			slog.Warn(fmt.Sprintf("%s channel handshake is null, sessionId = %s", channel.GetConfig().GetRoleName(), channel.GetSession().SessionId()))
			return nil
		}

		switch frame.Flag {
		case constant.FramePing:
			channel.SendPong()
		case constant.FramePong:
			channel.SendPong()
		case constant.FrameClose:
			channel.Close(constant.CLOSE1_PROTOCOL)
			p.OnCloseInternal(channel)
		case constant.FrameAlarm:
			//TODO
		case constant.FrameMessage:
			fallthrough
		case constant.FrameRequest:
			fallthrough
		case constant.FrameSubscribe:
			p.OnReceiveDo(channel, frame, false)
		case constant.FrameReply:
			p.OnReceiveDo(channel, frame, true)
		case constant.FrameReplyEnd:
			p.OnReceiveDo(channel, frame, true)
		default:
			channel.Close(constant.CLOSE2_PROTOCOL_ILLEGAL)
			p.OnCloseInternal(channel)
		}
	}
	return nil
}

func (p *Processor) OnReceiveDo(channel core.ChannelInternal, frame *message.Frame, reply bool) {
	var stm stream.StreamInternal
	var streamIndex = 0
	var streamTotal = 1
	if reply {
		stm = channel.GetStream(frame.Message.Sid)
	}
	if channel.GetConfig().GetFragmentHandler().AggrEnable() {
		var fragmentIdxStr = frame.Message.Meta.Get(constant.META_DATA_FRAGMENT_IDX)
		if fragmentIdxStr != "" {
			streamIndex, _ = strconv.Atoi(fragmentIdxStr)
			frameNew, _ := channel.GetConfig().GetFragmentHandler().AggrFragment(channel, streamIndex, frame.Message)

			if stm != nil {
				streamTotal, _ = strconv.Atoi(frame.Message.Meta.Get(constant.META_DATA_FRAGMENT_TOTAL))
			}
			if frameNew == nil {
				if stm != nil {
					stm.OnProgress(false, streamIndex, streamTotal)
				}
				return
			}
			frame.Message = frameNew
		}
	}

	//执行接收处理
	if reply {
		if stm != nil {
			stm.OnProgress(false, streamIndex, streamTotal)
		}
		channel.Retrieve(frame, stm)
	} else {
		p.OnMessage(channel, frame)
	}
}

func (p *Processor) OnOpen(channel core.ChannelInternal) {
	p.listener.OnOpen(channel.GetSession())
	channel.DoOpenFuture(true, nil)
}

func (p *Processor) OnMessage(channel core.ChannelInternal, message *message.Frame) {
	var err = p.listener.OnMessage(channel.GetSession(), message)
	if err != nil {
		p.OnError(channel, err)
	}
}

func (p *Processor) OnClose(channel core.ChannelInternal) {
	if channel.IsClosed() == 0 {
		p.OnCloseInternal(channel)
	}
}

func (p *Processor) OnCloseInternal(channel core.ChannelInternal) {
	p.listener.OnClose(channel.GetSession())
}

func (p *Processor) OnError(channel core.ChannelInternal, err error) {
	p.listener.OnError(channel.GetSession(), err)
}
